Context-aware dynamical decoupling

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Implement context-aware dynamical decoupling (DD) from [arXiv:2403.06852](https://arxiv.org/abs/2403.06852).

### Details and comments

When selecting decoupling sequences for an idle qubit, context-aware DD takes into account what actions are occuring on neighboring qubits. It ensures that neighboring sequences are mutually orthogonal (to avoid collisions, like in staggered DD) but also whether the qubit is a CX or ECR target/control spectator. This is nicely summarized in a figure of the paper, where the different wire colors denote different decoupling sequences:

<img width="1410" alt="image" src="https://github.com/user-attachments/assets/485a182c-1fc6-4ddb-ba37-aac0515e99b6">

This particularly leads to improvement if a lot of qubits are idle at the same time or if they are next to ECR/CX gates. One important category of circuits in this class are the layers in twirled circuits. This pass also works best if the ECR/CX gate times are the same for all connections, which is almost always the case on IBM devices.

Some simple experiments show that this pass performs better than the current standard in Qiskit, `PadDynamicaDecoupling` with an `X-X` sequence. Here's some basic benchmarks, see the paper for more: 

<img width="1912" alt="image" src="https://github.com/user-attachments/assets/1eb89b94-a956-46ac-adc1-f348e0e6b432">

This code is based off @kdk and @ajavadia 🙂

Cryoris committed:
Context aware DD

Co-authored-by: Kevin Krsulich <kevin.krsulich@ibm.com>
Co-authored-by: Ali Javadi <ajavadia@princeton.edu>

Cryoris committed:
add reno

Cryoris committed:
fix import

Cryoris committed:
fix test by using GenericBackendV2

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,785 @@
+    return closed_delay_blocks
+
+
+def _split_adjacent_block(
```
Any particular reason why this and all the functions below are not instance methods of the `ContextAwareDynamicalDecoupling` class?

Cryoris replied:
Simply because they don't depend on `self` and I wanted to keep the class lean, but looking at it again, the functions depend on the class context (e.g. `self.property_set`) so putting them inside the class would make sense 🙂 I'll update!

Cryoris replied:
Done in c6700ebb7.

Cryoris committed:
Merge branch 'main' into context-aware-dynamical-decoupling

Cryoris committed:
use class method instead of standalones

done as the functions use data already available in the class-context, so we can avoid redundant passing of variables

miamico commented:
Anything holding this up from being merged? Happy to lend a hand if needed to speed up the process @Cryoris @eliarbel

eliarbel commented on qiskit/transpiler/passes/__init__.py:
```diff
@@ -286,6 +286,7 @@
 from .scheduling import ASAPSchedule
 from .scheduling import DynamicalDecoupling
 from .scheduling import AlignMeasures
+from .scheduling import ContextAwareDynamicalDecoupling
```
Need to add the class also to the Scheduling section above in the documentation part so it'll be included in the docs.

Cryoris replied:
added in 13802fc36

eliarbel commented on test/python/transpiler/test_context_aware_dd.py:
```diff
@@ -0,0 +1,734 @@
+
+@ddt
+class TestContextAwareDD(QiskitTestCase):
+    """Test context-aware dynamical decoupling."""
```
For convenience I would put the link to the paper also here

Cryoris replied:
added in 13802fc36

eliarbel commented on test/python/transpiler/test_context_aware_dd.py:
```diff
@@ -0,0 +1,734 @@
+        circuit.barrier()
+        circuit.cx(1, 2)
+        circuit.cx(4, 5)
+        circuit.barrier()
```
The penultimate layer of Fig. 5 is missing here, namely cx(1,0) and cx(4,5). Is this intentional?

Cryoris replied:
Good catch! Though this is actually legacy code and no longer used in the testing 🙂 I'll remove the whole method.

Cryoris replied:
removed in 13802fc36

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+
+
+class ContextAwareDynamicalDecoupling(TransformationPass):
+    """Context-aware dynamical decoupling.
```
Since the summary in the Scheduling TOC in the docs is taken from here, I would make this a bit more informative otherwise it's just repeating the class name.

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+
+    This pass implements a context-aware dynamical decoupling (DD) [1], which ensures that
+
+        (1) simultaneously occuring DD sequences on device-adjacent qubits are mutually orthogonal, and
```
```suggestion
        (1) simultaneously occurring DD sequences on device-adjacent qubits are mutually orthogonal, and
```

Cryoris replied:
fixed in 13802fc36

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+            pulses on the neighboring target/control qubits.
+
+    The mutually orthogonal DD sequences are currently Walsh-Hadamard sequences, consisting of only
+    X gates. In some cases it might therefore be benefitial to use :class:`.PadDynamicalDecoupling`
```
```suggestion
    X gates. In some cases it might therefore be beneficial to use :class:`.PadDynamicalDecoupling`
```

Cryoris replied:
Whoops I didn't know this spelling is incorrect -- thanks!

Cryoris replied:
fixed in 13802fc36

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+
+                op.start_times += start_times
+
+        # 5) Replace each delay operation with it's individual DD sequence
```
```suggestion
        # 5) Replace each delay operation with its individual DD sequence
```

Cryoris replied:
fixed in 13802fc36

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+        # 5) Replace each delay operation with it's individual DD sequence
+        qubit_map = dict(enumerate(dag.qubits))
+        for delay in all_delays:
+            # replace it with it's stored replacement
```
```suggestion
            # replace it with its stored replacement
```

Cryoris replied:
fixed in 13802fc36

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+        elif order == 5:
+            spacing = [1 / 8, 1 / 8, 1 / 8, 1 / 4, 1 / 8, 1 / 8, 1 / 8]
+        else:
+            max_order = ContextAwareDynamicalDecoupling.MAX_ORDER
```
What scenarios would lead to asking for orders higher than 5? Worth explaining why 5 is the max order.

Cryoris replied:
Because I got to lazy to write the code that generalizes this and thought there's not frequently a device which has a qubit with more than 5 connections 😉 I'll add the code that does this generally

Cryoris replied:
Added in 13802fc36. This can now compute Walsh Hadamard sequences up to arbitrary orders.

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+    def _get_wire_coloring(self, dag: DAGCircuit, merged_delay: MultiDelayOp) -> dict[int, int]:
+        """Find a wire coloring for a multi-delay operation.
+
+        This function returns a dictionay that includes the coloring (as int) for the indices in the
```
```suggestion
        This function returns a dictionary that includes the coloring (as int) for the indices in the
```

Cryoris replied:
fixed in 13802fc36

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+            open_delay_block.events.append(delay_event)
+
+            if delay_event.type == EventType.BEGIN:  # TODO: this is never called
+                open_delay_block.active_qubits += set(delay_event.op_node.qargs)
```
This is branch is indeed not taken as `_update_delay_block` is only called with EventType.END. I would clean this then and put `assert delay_event.type != EventType.BEGIN`

Cryoris replied:
Removed entirely in 13802fc36

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+                # If so, add current event to that group.
+
+                if len(adjacent_open_delay_blocks) == 0:
+                    # Make a new delay
```
```suggestion
                    # Make a new delay block
```

Cryoris replied:
fixed in 13802fc36

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+                    new_block = _open_delay_block(delay_event)
+                    _combine_delay_blocks(adjacent_open_delay_blocks + [new_block])
+
+            if delay_event.type == EventType.END:
```
Just nitpicking, but I would make this an `else:` instead.

Cryoris replied:
done in 13802fc36

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+            if delay_event.type == EventType.END:
+                # If crossing a end edge, remove this qubit from the actively delaying qubits"
+                if len(adjacent_open_delay_blocks) != 1:
+                    raise RuntimeError("Closing edge w/o an open delay?")
```
There are several `RuntimeError` raised in the code but it's not clear if those are meant to be for the user or of the developer. This one in particular seems more like a developer check to me, so I would transform it to an `assert` statement. I would only raise exceptions for cases the users should act upon, e.g. in cases of invalid inputs (e.g. DAG/Target).

mtreinish replied:
I agree with Eli here about the intent of these exception not being clear or user actionable, but I don't think a runtime assertion is a good choice. 

Using `assert` is not really desirable either because the raised `AssertionError` is also not really actionable and typically ends up being even more cryptic, especially since there is little control over the exception raised. I'd argue either change the exceptions raised to a qiskit exception type and change the error message to indicate how a user can address the failure mode or just remove the check completely. If it's just an internal sanity check for debugging we should remove it and avoid the runtime overhead completely. In theory this is what you can do with asserts, because `python -O` will strip the asserts but in practice I've very rarely seen anyone use that flag.

jakelishman replied:
Fwiw, we have used `RuntimeError` for other places in Qiskit to represent an internal logic error, but the idea is that it should be entirely impossible for a user to trigger them (unless they're doing something they're not allowed to), or they're a flag to implementers that they've done something wrong.

That said, I believe that the majority of places it's used in Qiskit already are places where it has no runtime cost to do, because it's the "default" case of a lookup table or an `if/else` chain, or something like that, and we need to do _something_ other than just fall off the end of the chain silently.  I think quite a few of the uses here don't fall into those buckets, and should be replaced with proper errors - some of them are user actionable, and some are internal logic validation that has non-trivial overhead to check.

Cryoris replied:
I opted for the third variant: adding optional `logger.debug` statements, if the respective logging level is enabled. This code has been tested on dummy circuits and a variety of smaller hardware experiments, but I'd prefer keeping an easy way of printing debug info as long as the code is relatively new. Let me know if this is fine, otherwise I can also remove these!

eliarbel replied:
Yes, a logging approach is good

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+
+        for closed_delay in closed_delay_blocks:
+            if len(closed_delay.active_qubits) > 0:
+                raise RuntimeError("Failed remove active qubits on closed delays.")
```
Also here and a few lines above - are these developer checks or user errors?

Cryoris replied:
They are dev checks -- see comment above 🙂

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+            if len(closed_delay.active_qubits) > 0:
+                raise RuntimeError("Failed remove active qubits on closed delays.")
+
+            if validate:
```
There are several validation checks across the code. I suggest enabling and using them in `test_context_aware_dd.py`

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+                    # TODO this is a sanity check
+                    involved_ops = {op_map[(index, window)] for index in open_group}
+                    if len(involved_ops.difference(delay.ops)) > 0:
+                        raise RuntimeError("I cannot think")
```
You are thinking too classically ;-)
Make it an assertion or part of a validation block

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+
+    @staticmethod
+    def sort_key(event: DelayEvent) -> tuple(int, int):
+        """Sort events, first by time then by type ('end' events come for 'begin' events)."""
```
```suggestion
        """Sort events, first by time then by type ('end' events come before 'begin' events)."""
```

Cryoris replied:
fixed in 13802fc36

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+        q0: -██████---------  |  qubits q0,q1,q2 have adjacent delay
+        q1: ------███████---  |  operations, since the delay operations
+        q2: --█████████-----  |  all overlap
+        q3: -----------████-  -> this delay starts when delay on q2 end, so they have no overlap
```
```suggestion
        q3: -----------████-  -> this delay starts when delay on q2 ends, so they have no overlap
```

Cryoris replied:
done in 13802fc36

eliarbel reviewed (changes requested):
Thanks Julien, overall seems well written and documented. I didn't read the paper so the review is not about the actual algorithmic part of the DD method itself. I suggest getting input from Kevin or Ali if you want another pair of eyes on this aspect. 
I've left line comments, most of which are minor but still require addressing. 
In terms of performence - do you have a sense about the overhead of the block collection and manipulation (e.g. `_collect_adjacenet_delay_blocks` )?

ElePT commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+            target: The :class:`.Target` of the device to run the circuit.
+            min_joinable_duration: Do not join delays with others below this duration. This
+                can be useful, e.g. if a big delay block would be interrupted and split into
+                smaller blocks due to a very short, adjacent delay.
```
I would find it very helpful if the documentation mentioned something that could be linked easier to "minimal delay time" and what units are used, something like (I still find it a bit clunky):

```suggestion
            min_joinable_duration: Minimal duration to join delays (in dt units). Do not join delays with
                others below this duration . This can be useful, e.g. if a big delay block would be interrupted
                and split into smaller blocks due to a very short, adjacent delay.
```

Cryoris replied:
Done in [6c6f6d7](https://github.com/Qiskit/qiskit/pull/12825/commits/6c6f6d7ff633c1246a3ac08dce07fb922011f8a2)

ElePT commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+                can be useful, e.g. if a big delay block would be interrupted and split into
+                smaller blocks due to a very short, adjacent delay.
+            skip_reset_qubits: Skip initial delays and delays after a reset.
+            skip_dd_threshold: Skip dynamical decoupling on an idle qubit, if the duration of
```
In what units is `skip_dd_threshold`? I have noticed that `min_joinable_duration` is an int but this one is a float, so maybe it's not dt?

Cryoris replied:
It's in the fraction of the idle window 🙂 
> Skip dynamical decoupling on an idle qubit, if the duration of
                the decoupling sequence exceeds this fraction of the idle window.
                
I'll add a clarifying comment, maybe something like
```
For example, to skip DD if the DD sequence would take up
more than 95% of the idle time, set this to 0.95. A value of 
1.0 means that DD is always used if it fits into the sequence.
```

Cryoris replied:
Done in [6c6f6d7](https://github.com/Qiskit/qiskit/pull/12825/commits/6c6f6d7ff633c1246a3ac08dce07fb922011f8a2)

ElePT commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,756 @@
+            pulse_alignment: The hardware constraints for gate timing allocation.
+                This is usually provided om ``backend.configuration().timing_constraints``.
+                If provided, the delay length, i.e. ``spacing``, is implicitly adjusted to
+                satisfy this constraint. If ``None``, this is extracted from the ``target``.
```
Same comment about units

Cryoris replied:
Done in [6c6f6d7](https://github.com/Qiskit/qiskit/pull/12825/commits/6c6f6d7ff633c1246a3ac08dce07fb922011f8a2)

ElePT reviewed (commented):
Thanks @Cryoris, as you know I have ran this pass and noticed a few painpoints I wanted to point out. In addition to the documentation suggestions, I have noticed that `min_joinable_duration` is a key parameter to avoid 1q gate explosion in some circuits, I wonder if you had considered setting it no a non-0 default value?

ElePT reviewed (commented):
Another comment, I believe there is a bug in the instruction duration assignation, because the timeline drawer fails after running this pass complaining about missing gate durations:

```
Traceback (most recent call last):
  File "/Users/ept/qiskit_workspace/qiskit/debug_dd.py", line 84, in <module>
    timeline_drawer(bound_scheduled_circuit, filename="out.png")
  File "/Users/ept/qiskit_workspace/qiskit/qiskit/utils/deprecation.py", line 183, in wrapper
    return func(*args, **kwargs)
  File "/Users/ept/qiskit_workspace/qiskit/qiskit/utils/deprecation.py", line 183, in wrapper
    return func(*args, **kwargs)
  File "/Users/ept/qiskit_workspace/qiskit/qiskit/visualization/timeline/interface.py", line 379, in draw
    canvas.load_program(program=program)
  File "/Users/ept/qiskit_workspace/qiskit/qiskit/visualization/timeline/core.py", line 193, in load_program
    for datum in gen(gate_source, formatter=self.formatter):
  File "/Users/ept/qiskit_workspace/qiskit/qiskit/visualization/timeline/generators.py", line 158, in gen_sched_gate
    if gate.duration > 0:
TypeError: '>' not supported between instances of 'NoneType' and 'int'
```

Cryoris committed:
Eli's review comments

Cryoris committed:
Merge branch 'main' into context-aware-dynamical-decoupling

eliarbel reviewed (commented):
Thanks @Cryoris for applying all the changes. This looks good and I have no further comments. There are still some comments from @ElePT.

eliarbel reviewed (dismissed):
Fine from my side. Please address the remaining comments from @ElePT

Cryoris committed:
Merge branch 'main' into context-aware-dynamical-decoupling

Cryoris commented:
https://github.com/Qiskit/qiskit/pull/12825#pullrequestreview-2404327735 is a very good catch -- the newly added gates that implement the DD sequences didn't have a `duration` attached, since they weren't present when the scheduling pass was run. I'll add a test to check for this 🙂

Cryoris committed:
Elena's review comments

* add details about time units
* add gate.durations for inserted sequences

Cryoris committed:
smarter default for min_joinable_durations

this needs some benchmarking on real devices

Cryoris committed:
lint + undo unwanted changes

Cryoris commented:
After discussion with @ElePT I also updated the default behavior of `min_joinable_duration`: On realistic devices, the 2q gate lengths vary a little -- yet might not want to insert DD sequences for two concurrent 2q gates, just because they naturally differ slightly. The new default is therefore that we only add DD sequences for idle times beyond `2 * (max_2q_duration - min_2q_duration)`. This needs some benchmarking on real devices before merging.

mtreinish commented:
> [#12825 (review)](https://github.com/Qiskit/qiskit/pull/12825#pullrequestreview-2404327735) is a very good catch -- the newly added gates that implement the DD sequences didn't have a `duration` attached, since they weren't present when the scheduling pass was run. I'll add a test to check for this 🙂

Just so you know I'm working on removing per instruction durations in #13506 so we probably don't need to change this now. It would have been important for 1.3, but in case 2.0 we just making extra work for both of us.

Cryoris committed:
don't use deprecated op.duration

Cryoris commented:
I removed any setting and usage of `Instruction.duration` in 21c4ea0, so hopefully you don't have to update the context-aware DD in #13506 🙂 That means the timeline drawer will fail now, but I assume this will be fixed in your PR 🙂

eliarbel committed:
Merge branch 'main' into context-aware-dynamical-decoupling

eliarbel commented:
> After discussion with @ElePT I also updated the default behavior of `min_joinable_duration`: On realistic devices, the 2q gate lengths vary a little -- yet might not want to insert DD sequences for two concurrent 2q gates, just because they naturally differ slightly. The new default is therefore that we only add DD sequences for idle times beyond `2 * (max_2q_duration - min_2q_duration)`. This needs some benchmarking on real devices before merging.

Hi @Cryoris are you planning to benchmark the new `min_joinable_duration` default on real devices before the 2.1 release?

Cryoris committed:
Merge branch 'main' into context-aware-dynamical-decoupling

Cryoris committed:
fix tests after merge

Cryoris committed:
:wqMerge branch 'context-aware-dynamical-decoupling' of github.com:Cryoris/qiskit-terra into context-aware-dynamical-decoupling

Cryoris commented:
Yep, it's good from my side now 🙂 The number of X gates is rather large though, per design of the algorithm, I think that was just surprising to see. 

E.g. for an N-local circuit with linear entanglement on `n` qubits (not that someone would run this), the number of X gates goes with `O(n^2)` instead of `O(n)` for "normal" DD. But the results of CA-DD look promising on the eagle devices and the overhead can be controlled with the `min_joinable_duration` parameter.

Cryoris committed:
handle `dt` none case

Cryoris committed:
Fix tests after min duration default update

eliarbel commented:
Overall this looks in good shape to me both from code and testing perspectives. I haven't reviewed the algorithm in depth against the paper, but I ran several examples and the results look OK. Still, I'd like to look into some of the algorithmic aspects a bit more. In the meantime, I have a couple questions and possible suggestions, both not major: 

The rationale behind the default value for `min_joinable_duration` is not so clear to me. Also, it doesn't feel very intuitive as a knob for controlling the amount of DD inserted into the circuit. Is there some underlying intuition behind the choice, something that could be added to the docstring to help clarify this a bit?

'cz' is included in the calculation of the 2q gate length variance in `_gate_length_variance`. However, the logic for handling spectator qubits does not account for `CZGate`s. Was this intentional, perhaps due to the differences in crosstalk levels between Heron and Eagle devices? If so, this introduces a technology assumption in the implementation, isn't it? I'm wondering if it would make sense to provide a way for users to opt in or out of spectator qubit analysis for specific gate types?

Cryoris commented:
Re: `min_joinable_duration` -- this shouldn't set the amount of DD inserted, but rather control how adjacent idle blocks are joined. It's a knob that lets you trade-off between potentially many short sequences or longer blocks. The default value is picked to ensure that a layer of CX/ECR will not have a lot of short DD sequences just due to the difference in duration. But this is not discussed in the paper afaik and I'm happy to change this if you think there's a more reasonable choice 🙂 

Re: CZ -- The paper only investigates CX/ECR. CZ would probably require a different spectator qubit handling, but this is not explained in the paper and hence not supported in the code. We can add an argument that allows to turn off the spectator-specific logic for CX/ECR.

eliarbel commented:
> Re: min_joinable_duration -- this shouldn't set the amount of DD inserted, but rather control how adjacent idle blocks are joined. It's a knob that lets you trade-off between potentially many short sequences or longer blocks. The default value is picked to ensure that a layer of CX/ECR will not have a lot of short DD sequences just due to the difference in duration. But this is not discussed in the paper afaik and I'm happy to change this if you think there's a more reasonable choice 🙂

The reason I treated this is as a knob to control the amount of DD is because of this check: 
```python
if (
    node.op.name == "delay"
    and not is_after_reset(node)
    and self._duration(node, qubit_map) > self._min_joinable_duration
   )
```
in `_get_sorted_delays`.  It skips durations shorter than `min_joinable_duration` when collecting eligible delays, effectively becoming a filter of idle blocks based on their lengths. This is because shorter delays are not just omitted from adjacency blocks and have their own DD sequences, they are skipped altogether. If we want a truly `min_joinable_duration` behavior in the "joinable" sense, we'll probably want to handle the parameter in `_collect_adjacent_delay_blocks`.  As for the value default value itself, I see the reasoning for layers so I think it's OK. But now I think that a name that better reflects how the logic is implemented currently is `min_duration` 😄

Cryoris commented:
Fair point, let's rename it 👍🏻

Cryoris committed:
Merge branch 'main' into context-aware-dynamical-decoupling

Cryoris committed:
rename min_(joinable)_duration

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,875 @@
+
+    Example::
+
+        from qiskit.circuit.library import QFTGate
```
```suggestion
        from qiskit.circuit import QuantumCircuit
        from qiskit.circuit.library import QFTGate
```

eliarbel commented on qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py:
```diff
@@ -0,0 +1,875 @@
+        q1: ------███████---  |  operations, since the delay operations
+        q2: --█████████-----  |  all overlap
+        q3: -----------████-  -> this delay starts when delay on q2 ends, so they have no overlap
+        q4: ----████--------  -> this clearly has no overlap with someting else
```
```suggestion
        q4: ----████--------  -> this clearly has no overlap with something else
```

eliarbel reviewed (changes requested):
Thanks @Cryoris for the patience with the final review. I only have 2 minor docs comments, otherwise this looks very good!

Cryoris committed:
Apply suggestions from code review

Thanks Eli!

Co-authored-by: Eli Arbel <46826214+eliarbel@users.noreply.github.com>

eliarbel reviewed (approved):
Great work - good to go! 🚀
